src/main/webapp/scripts/tests/popular-test.js

```diff
@@ -0,0 +1,18 @@
+import { addRegionOptions } from '../popular.js';
+
+const mockGetRegionsResponse = {
+  items: [
+    { snippet: { gl: 'EG', name: 'Example'}}
+  ]
+}
+
+QUnit.test('region menu display',  function(assert) {
+  const done = assert.async();
+  const getRegions = sinon.fake.returns(
+    Promise.resolve(mockGetRegionsResponse)
+  );
```

Comment:
This will not work because `getRegions` is in a different scope from this function. One way to mock getRegions is to import its reference from popular.js:

```
import * as popular from '../popular.js';
.
.
.
let { addRegionOptions, getRegions } = popular;
getRegions = sinon.fake ...
```

You'll also need to export `getRegions` from popular.js

Reply:
It's still not working :/ did I interpret wrong? I am still getting no error message but it is using the real function instead of the fake. I also tried exporting getRegions in popular-test and `import { getRegions } from './tests/popular-test'` in the popular.js file but that caused an error because the functions have the same name. How can I specify when I want to use the getRegions in the original file vs the getRegions fake from the test file...

Reply:
Sorry my instructions were not right. Your export in popular.js should be a default export:

`export default {addRegionOptions, getRegions}`

and the import in this file should be 

`import popular from '../popular.js';`

you should use sinon.replace to mock getRegions

```
QUnit.test('region menu display',  function(assert) {
  const done = assert.async();
  const getRegions = sinon.fake.returns(
    Promise.resolve(mockGetRegionsResponse)
  );
  sinon.replace(popular, 'getRegions', getRegions);
  popular.addRegionOptions().then(() => {
    assert.dom('option').hasText('Example');
    done();
  });
});
```

Reply:
Sorry, my instructions were not right. This works on my machine:
```
//popular.js
import { getRegions } from './region-selector.js';
let popular;

function addRegionOptions() {
  return loadApi().then(() => popular.getRegions()) // call popular.getRegions() here
  .then((response) => {
    for(const region of response.items) {
      addOptionToSelectList(region.snippet.gl, region.snippet.name, 'region-select');
    }
  });
}   
  
export default popular = {addRegionOptions, getRegions} 
```

```import popular from '../popular.js';

const mockGetRegionsResponse = {
  items: [
    { snippet: { gl: 'EG', name: 'Example'}}
  ]
}   
      
QUnit.test('region menu display',  function(assert) {
  const done = assert.async();
  const getRegions = sinon.fake.returns(
    Promise.resolve(mockGetRegionsResponse)
  );
  sinon.replace(popular, 'getRegions', getRegions);
  popular.addRegionOptions().then(() => {
    assert.dom('option').hasText('Example');
    done();
  });
});```

Reply:
`let popular` is necessary because ES6 imports are readonly so you have to use `let` to be able to overwrite the import.

Normally ES6 code is transpiled to ES5 (to support older browsers) where the original instructions should work.

Reply:
Sorry to cut in -- what's the rationale behind exporting the dependency `export default popular = {addRegionOptions, getRegions}` in popular.js, rather than importing `getRegions` from `./region-selector.js` in popular-test.js?

Reply:
Exporting the `getRegions` dependency from popular.js is needed order to get the same reference to `getRegions` that is being used in `popular.js`.

Note that in `popular.js` we're calling `loadApi().then(popular.getRegions)` instead of `loadApi().then(getRegions)` to ensure that the same reference as the export is used.

See the stackoverflow link above for more details.

Reply:
Thank you! I wasn't clear on the reference of function previously but now I get it.